handlers/analytics/log_processor.go

```diff
@@ -1,212 +1,162 @@
-// - on every analytics heartbeat event, process log which means updating the internal structure with the log data
-// - every SendMetricsInterval, send metrics to the Prometheus (Victoria Metrics) DB
-// Note that it sends the metrics in the plaintext format, this could be changed to sending data in the binary format,
-// but plaintext seems to be efficient enough.
+// Start starts LogProcessor which sends events to Kafka in batches.
 func (lp *LogProcessor) Start(ch chan LogData) {
-	t := time.NewTicker(SendMetricsInterval)
-
+	t := time.NewTicker(KafkaBatchInterval)
 	go func() {
 		for {
 			select {
 			case d := <-ch:
 				lp.processLog(d)
 			case <-t.C:
-				lp.sendMetrics()
+				lp.sendEvents()
 			}
 		}
 	}()
 }
 
 func (p *LogProcessor) processLog(d LogData) {
-	var k = labelsKey{
-		playbackID: d.PlaybackID,
-		browser:    d.Browser,
-		deviceType: d.DeviceType,
-		country:    d.Country,
-		userID:     d.UserID,
-	}
-
-	bySessionID, ok := p.logs[k]
-	if !ok {
-		p.logs[k] = make(map[string]metricValue)
-		bySessionID = p.logs[k]
-	}
-
-	totalMs := d.PlaytimeMs + d.BufferMs
-	var rebufferRate float64
-	if totalMs > 0 {
-		rebufferRate = float64(d.BufferMs) / float64(totalMs)
-	}
-	var errorRatio float64
-	if d.Errors > 0 {
-		errorRatio = 1
-	}
-	mv := bySessionID[d.SessionID]
-	bySessionID[d.SessionID] = metricValue{
-		count:           mv.count + 1,
-		sumRebufferRate: mv.sumRebufferRate + rebufferRate,
-		sumErrorRatio:   mv.sumErrorRatio + errorRatio,
-	}
+	p.logs = append(p.logs, d)
 }
 
-func (p *LogProcessor) sendMetrics() {
+func (p *LogProcessor) sendEvents() {
 	if len(p.logs) > 0 {
 		glog.Info("sending analytics logs")
 	} else {
 		glog.V(6).Info("no analytics logs, skip sending")
 		return
 	}
 
-	// convert values in the Prometheus format
-	var metrics strings.Builder
-	now := time.Now().UnixMilli()
-	for k, v := range p.logs {
-		metrics.WriteString(p.toViewCountMetric(k, v, now))
-		metrics.WriteString(p.toRebufferRatioMetric(k, v, now))
-		metrics.WriteString(p.toErrorRateMetric(k, v, now))
-	}
-
-	// send data
-	err := p.sendMetricsString(metrics.String())
-	if err != nil {
-		glog.Errorf("failed to send analytics logs, err=%v", err)
-	}
-
-	// clear map
-	p.logs = make(map[labelsKey]map[string]metricValue)
-}
-
-func (p *LogProcessor) toViewCountMetric(k labelsKey, v map[string]metricValue, nowMs int64) string {
-	value := fmt.Sprintf("%d", len(v))
-	return p.toMetric(k, "view_count", value, nowMs)
-}
-
-func (p *LogProcessor) toRebufferRatioMetric(k labelsKey, v map[string]metricValue, nowMs int64) string {
-	var count int
-	var sumRebufferRate float64
-	for _, mv := range v {
-		if mv.count > 0 {
-			count += 1
-			sumRebufferRate += mv.sumRebufferRate / float64(mv.count)
+	var msgs []kafka.Message
+	for _, d := range p.logs {
+		key := []byte(d.SessionID)
+		value, err := json.Marshal(d)
+		if err != nil {
+			glog.Errorf("invalid analytics log event, cannot sent to Kafka, err=%v", err)
```

Comment:
Can we get an identifying field in here too? e.g SessionID

Reply:
Added sessionID